FhirToDataLake/src/Microsoft.Health.AnalyticsConnector.Core/Jobs/FhirToDataLakeOrchestratorJob.cs

```diff
@@ -212,7 +228,159 @@ await foreach (FhirToDataLakeProcessingJobInputData input in inputs.WithCancella
+                ProcessedDataSizeInTotal = _jobStatus.ProcessedDataSizeInTotal,
+            };
+        }
+
+        private async Task<FhirToDataLakeOrchestratorJobStatus> InitializeJobStatusFromTableAsync(CancellationToken cancellationToken)
+        {
+            _jobStatusEntity = await _metadataStore.GetJobStatusAsync(_jobInfo.QueueType, _jobInfo.GroupId, _jobInfo.Id, cancellationToken);
+
+            if (_jobStatusEntity == null)
+            {
+                // Orchestrator job status entity is not exist, will create a new one.
+                var newJobStatusEntity = new JobStatusEntity()
+                {
+                    PartitionKey = TableKeyProvider.JobStatusPartitionKey(_jobInfo.QueueType, _jobInfo.GroupId),
+                    RowKey = TableKeyProvider.JobStatusRowKey(_jobInfo.GroupId, _jobInfo.Id),
+                    JobType = JobType.Orchestrator,
+                    GroupId = _jobInfo.GroupId,
+                    JobStatus = JsonConvert.SerializeObject(new FhirToDataLakeOrchestratorJobResult()),
+                };
+
+                if (!await _metadataStore.TryAddEntityAsync(newJobStatusEntity, cancellationToken))
+                {
+                    _logger.LogError("Initialize job status failed in orchestrator job {0}.", _jobInfo.Id);
+                    throw new SynapsePipelineInternalException($"Initialize job status failed in orchestrator job {_jobInfo.Id}.");
+                }
+
+                // Update Etag of job status entity.
+                _jobStatusEntity = await _metadataStore.GetJobStatusAsync(_jobInfo.QueueType, _jobInfo.GroupId, _jobInfo.Id, cancellationToken);
+            }
+
+            return JsonConvert.DeserializeObject<FhirToDataLakeOrchestratorJobStatus>(_jobStatusEntity.JobStatus);
+        }
+
+        private async Task PersistJobStatusAsync(IProgress<string> progress, CancellationToken cancellationToken)
+        {
+            // Usage of result will be deprecated from V5.
+            if (_inputData.JobVersion <= JobVersion.V4)
+            {
+                UpdateJobResultFromStatus();
+                progress.Report(JsonConvert.SerializeObject(_result));
+            }
+
+            if (_inputData.JobVersion >= JobVersion.V4)
+            {
+                // Update job status on meta table from V4 job version.
+                _jobStatusEntity.JobStatus = JsonConvert.SerializeObject(_jobStatus);
+
+                if (!await _metadataStore.TryUpdateEntityAsync(_jobStatusEntity))
+                {
+                    _logger.LogError("Update orchestrator job status failed in job {0}.", _jobInfo.Id);
+                    throw new SynapsePipelineInternalException($"Update orchestrator job status failed in job  {_jobInfo.Id}.");
+                }
+
+                _jobStatusEntity = await _metadataStore.GetJobStatusAsync(_jobInfo.QueueType, _jobInfo.GroupId, _jobInfo.Id, cancellationToken);
```

Comment:
what is the consideration to get _jobStatusEntity when persist job status. seems the _jobStatusEntity only is used when return the result. can we only get it before return the result?

Reply:
We need to update entity Etag here.